### PR TITLE
Rewrite bash scripts into Python interfaces 

### PR DESCRIPTION
**Pull request by rinapch**
Changelog: 
- Rewrite 4 main .sh scripts needed for experiment reproduction to Python 
- Switch project to poetry 
- Remove hardcode from get_dataset function, now you just need to provide the name and the load method 

Questions for discussion:
- The main role of the 2 bash scripts is to just set up some variables names and run a Python subprocess with the specified variables. We can just call the Python subprocess directly, however, with a trade-off that we will need to specify a lot of the variables by hand in the CLI 
- Should I also convert the training script for LoRA (the get the final model finetuned on selected data)? I didn't use as we have our own training script, but may be useful?

### CODE REVIEW

File: less/data_selection/matching.py [hunk 1]
```diff
@@ -44,23 +36,26 @@ def calculate_influence_score(training_info: torch.Tensor, validation_info: torc
         training_info, validation_info.transpose(0, 1))
     return influence_scores
 
+checkpoints = [int(i) for i in args.ckpts.split(" ")]
 
+checkpoint_weights = [float(i) for i in args.checkpoint_weights.split(" ")]
 # renormalize the checkpoint weights
```
Comment (mishaevtikhiev) [inline less/data_selection/matching.py:42]: I'm not sure how this matches with line 14, can you please explain?

File: less/scripts/get_info/grad/get_eval_lora_grads.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import os
+import argparse
+import subprocess
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--task", type=str, required=True, help="Task name (e.g. tydiqa, mmlu), will be used to store the gradients")
+    parser.add_argument("--data_dir", type=str, required=True, help="Path to data directory, can also be a full path or a HF repo name") 
+    parser.add_argument("--val_task_load_method", default=None,type=str, required=False, help="The method to load the validation data, can be 'hf', 'local_hf', 'local_json'")
```
Comment (mishaevtikhiev) [inline less/scripts/get_info/grad/get_eval_lora_grads.py]: Shouldn't this one be required? Looks like with None the script will fail at the dataset loading stage

File: less/scripts/get_info/grad/get_train_lora_grads.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+        # Build command
+        cmd = [
+            "python3", "-m", "less.data_selection.get_info",
+            "--train_file", args.train_file,
+            "--info_type", "grads",
+            "--model_path", model_checkpoint_path,
+            "--output_path", output_path,
+            "--gradient_projection_dimension", str(args.dims),
+            "--gradient_type", "adam"
```
Comment (mishaevtikhiev) [inline less/scripts/get_info/grad/get_train_lora_grads.py:36]: Sanity check question: Adam and AdamW provide the same type of gradient data, am I right?

File: less/scripts/train/warmup_lora_train.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    "--lora_alpha", "512",
+    "--lora_dropout", "0.1",
+    "--lora_target_modules", "q_proj", "k_proj", "v_proj", "o_proj",
+    "--learning_rate", "2e-05",
+    "--per_device_train_batch_size", "1",
+    "--gradient_accumulation_steps", "32"
+]
+
+    # Add FSDP config for large models
+    if model_name_or_path == "meta-llama/Llama-2-13b-hf":
```
Comment (mishaevtikhiev) [inline less/scripts/train/warmup_lora_train.py:45]: This is the original code from their .sh script, right?

**Review by mishaevtikhiev — APPROVED** [review]

File: less/data_selection/matching.py (hunk 1 above)
Comment (rinapch) [reply less/data_selection/matching.py:42]: If I understand your question correctly, checkpoint_weights has the same logic as checkpoints - you pass them as one string with spaces like so – `'1.6877e-05 1.2859e-05 7.7030e-06 2.5616e-06'`, to get an actual list of floats you need to parse this string first 
Or did you mean something else?

File: less/scripts/get_info/grad/get_eval_lora_grads.py (hunk 2 above)
Comment (rinapch) [reply less/scripts/get_info/grad/get_eval_lora_grads.py]: You're right, with None it would fail (fixed it). I made it optional before because for hardcoded initial datasets like tydiqa we don't need to specify it, but realistically we will only run our own datasets, so it's required now

File: less/data_selection/matching.py (hunk 1 above)
Comment (mishaevtikhiev) [reply less/data_selection/matching.py:42]: It was something else (I wasn't sure about checkpoint normalization, but misclicked at the line number), but after looking once again at the code I figured out the answer, so nevermind :)

File: less/scripts/get_info/grad/get_train_lora_grads.py (hunk 3 above)
Comment (rinapch) [reply less/scripts/get_info/grad/get_train_lora_grads.py:36]: In this repo in particular AdamW is used across all computations (despite this variable being just "Adam") -- it's used for warming up the model and for getting the gradients for the train data. For obtaining the gradients of eval data, gradient_type is set to "sgd" 


Here is the difference in computation: 
- SGD 
```
def obtain_gradients(model, batch):
    """ obtain gradients. """
    loss = model(**batch).loss
    loss.backward()
    vectorized_grads = torch.cat(
        [p.grad.view(-1) for p in model.parameters() if p.grad is not None])
    return vectorized_grads
``` 

- Adam(W) 
```
def prepare_optimizer_state(model, optimizer_state, device):
    names = [n for n, p in model.named_parameters() if p.requires_grad]
    avg = torch.cat([optimizer_state[n]["exp_avg"].view(-1) for n in names])
    avg_sq = torch.cat([optimizer_state[n]["exp_avg_sq"].view(-1)
                       for n in names])
    avg = avg.to(device)
    avg_sq = avg_sq.to(device)
    return avg, avg_sq


def obtain_gradients_with_adam(model, batch, avg, avg_sq):
    """ obtain gradients with adam optimizer states. """
    beta1 = 0.9
    beta2 = 0.999
    eps = 1e-08

    loss = model(**batch).loss
    loss.backward()

    vectorized_grads = torch.cat(
        [p.grad.view(-1) for n, p in model.named_parameters() if p.grad is not None])

    updated_avg = beta1 * avg + (1 - beta1) * vectorized_grads
    updated_avg_sq = beta2 * avg_sq + (1 - beta2) * vectorized_grads ** 2
    vectorized_grads = updated_avg / torch.sqrt(updated_avg_sq + eps)

    return vectorized_grads
 ```
 
 Looking at this, I think in principle this code would work with a normal Adam, but the gradient values produces by Adam and AdamW would be different

File: less/scripts/train/warmup_lora_train.py (hunk 4 above)
Comment (rinapch) [reply less/scripts/train/warmup_lora_train.py:45]: Yes, it seems like models larger than llama-7B would not fit into their GPUs (relatable) so they needed to enable FSDP for larger experiments

**Comment by kajigor** [comment]
Closed due to archival
